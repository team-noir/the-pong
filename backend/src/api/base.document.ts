import { SwaggerModule, DocumentBuilder } from "@nestjs/swagger";

export class BaseAPIDocumentation {
	public builder = new DocumentBuilder();

	public initializeOptions() {
		return this.builder
			.setTitle('The pong')
			.setDescription('The cats API description')
			.setVersion('1.0')
			.addTag('pong')
			.build();
	}
}
