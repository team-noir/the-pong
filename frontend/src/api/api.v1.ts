import axios from 'axios';
import { ChannelType } from 'types/channelType';
import { UserType } from 'types/userType';

export const API_PREFIX = `/api/v1`;

axios.defaults.baseURL = API_PREFIX;
axios.defaults.withCredentials = true;

const axiosWithInterceptors = axios.create();
axiosWithInterceptors.defaults.baseURL = API_PREFIX;
axiosWithInterceptors.defaults.withCredentials = true;

axiosWithInterceptors.interceptors.response.use(
  (response) => response,
  (error) => {
    if (error.response.status === 401) {
      window.location.href = '/login';
    }
    return Promise.reject(error);
  }
);

export const getHealthCheck = async () => {
  const res = await axios.get(`/health-check`);
  return res;
};

/** Login */

export const API_LOGIN_FT = `${API_PREFIX}/auth/42`;

export const postLogout = async () => {
  const res = await axiosWithInterceptors.post(`/auth/logout`);
  if (res.status !== 204) {
    throw new Error(res.statusText);
  }
  return res;
};

/** User */

export const getWhoami = async (): Promise<UserType> => {
  let whoamiAxios = axiosWithInterceptors;
  if (window.document.referrer === `${window.location.origin}/login`) {
    whoamiAxios = axios;
  }
  const res = await whoamiAxios.get(`/my/whoami`);
  if (res.status !== 200) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export interface ProfileUserType {
  id: number;
  nickname: string;
  rank: number;
  achievements: [];
  games: [];
  isFollowedByMyself: boolean;
  isBlockedByMyself: boolean;
}

export const getUser = async (userId: string): Promise<ProfileUserType> => {
  const res = await axiosWithInterceptors.get(`/users/${userId}`);
  if (res.status !== 200) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const getUsers = async (
  q: string,
  page = 1,
  per_page = 30
): Promise<UserType[]> => {
  const res = await axios.get(`/users`, {
    params: { q, page, per_page },
  });
  if (res.status !== 200) {
    throw new Error(res.statusText);
  }
  return res.data;
};

/** Setting */

export const patchMyProfile = async (nickname: string): Promise<UserType> => {
  const res = await axiosWithInterceptors.patch(`/my/settings`, { nickname });
  if (res.status !== 200) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const PostMyProfileImage = async (imageFile: File) => {
  const formData = new FormData();
  formData.append('file', imageFile);
  const res = await axiosWithInterceptors.post(`/my/profile-image`, formData, {
    headers: {
      'Content-Type': 'multipart/form-data',
    },
  });
  if (res.status !== 204) {
    throw new Error(res.statusText);
  }
  return res;
};

export const getMy2fa = async () => {
  const res = await axiosWithInterceptors.get(`/my/2fa`);
  if (res.status !== 200) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const deleteMy2fa = async () => {
  const res = await axiosWithInterceptors.delete(`/my/2fa`);
  if (res.status !== 204) {
    throw new Error(res.statusText);
  }
  return res;
};

/** Block */

export const getMyBlocks = async () => {
  const res = await axiosWithInterceptors.get(`/my/blocks`);
  if (res.status !== 200) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const deleteMyBlocks = async (userId: number) => {
  const res = await axiosWithInterceptors.delete(`/my/blocks/${userId}`);
  if (res.status !== 204) {
    throw new Error(res.statusText);
  }
  return res;
};

export const putMyBlocks = async (userId: number) => {
  const res = await axiosWithInterceptors.put(`/my/blocks/${userId}`);
  if (res.status !== 204) {
    throw new Error(res.statusText);
  }
  return res;
};

/** Follow */

export const getMyFollowing = async () => {
  const res = await axiosWithInterceptors.get(`/my/following`);
  if (res.status !== 200) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const deleteMyFollowing = async (userId: number) => {
  const res = await axiosWithInterceptors.delete(`/my/following/${userId}`);
  if (res.status !== 204) {
    throw new Error(res.statusText);
  }
  return res;
};

export const putMyFollowing = async (userId: number) => {
  const res = await axiosWithInterceptors.put(`/my/following/${userId}`);
  if (res.status !== 204) {
    throw new Error(res.statusText);
  }
  return res;
};

/** Channel */

interface getChannelsParams {
  enter?: string;
  kind?: string[];
}

export const getChannels = async ({ enter, kind }: getChannelsParams) => {
  const res = await axiosWithInterceptors.get(`/channels/`, {
    params: { enter, kind },
  });
  if (res.status !== 200) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export interface ChannelFormType {
  title: string;
  isPrivate?: boolean;
  password?: string;
}

export const postNewChannel = async (
  channelForm: ChannelFormType
): Promise<ChannelType> => {
  const res = await axios.post(`/channels`, channelForm);
  if (res.status !== 200) {
    throw new Error(res.statusText);
  }
  return res.data;
};
