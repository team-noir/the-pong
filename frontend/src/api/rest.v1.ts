import { StatusCodes } from 'http-status-codes';
import axios, { AxiosResponse } from 'axios';
import {
  UserType,
  AchievementType,
  GameType,
  GameHistoryType,
  ChannelType,
  ChannelUserRoleType,
  ChannelUserStatusType,
  ChannelFormType,
  MessageType,
  ListWithPagingType,
  PageParamsType,
} from 'types';

export const API_PREFIX = `${process.env.REACT_APP_API_URL}/api/v1`;

axios.defaults.baseURL = API_PREFIX;
axios.defaults.withCredentials = true;

const axiosWithInterceptors = axios.create();
axiosWithInterceptors.defaults.baseURL = API_PREFIX;
axiosWithInterceptors.defaults.withCredentials = true;

axiosWithInterceptors.interceptors.response.use(
  (response) => response,
  (error) => {
    if (error.response.status === StatusCodes.UNAUTHORIZED) {
      window.location.href = '/login';
    }
    return Promise.reject(error);
  }
);

export const healthCheck = async () => {
  const res = await axios.get(`/health-check`);
  return res;
};

/** Auth */

export const API_LOGIN_FT = `${API_PREFIX}/auth/42`;

export const anonymousLogin = async () => {
  const res = await axiosWithInterceptors.post(`/auth/login/anonymous`);
  if (res.status !== StatusCodes.OK) {
    throw new Error(res.statusText);
  }
  return res;
};

export const logout = async () => {
  const res = await axiosWithInterceptors.post(`/auth/logout`);
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

export const get2faCode = async (): Promise<{ qr: string; key: string }> => {
  const res = await axiosWithInterceptors.get(`/auth/2fa`);
  if (res.status !== StatusCodes.OK) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const verify2fa = async (code: string) => {
  const res = await axiosWithInterceptors.post(`/auth/2fa`, { code });
  if (res.status !== StatusCodes.ACCEPTED) {
    throw new Error(res.statusText);
  }
  return res;
};

export const delete2fa = async () => {
  const res = await axiosWithInterceptors.delete(`/auth/2fa`);
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

/** User */

export const whoami = async (): Promise<UserType> => {
  let whoamiAxios = axiosWithInterceptors;
  if (window.document.referrer === `${window.location.origin}/login`) {
    whoamiAxios = axios;
  }

  const res = await whoamiAxios.get(`/my/whoami`);
  if (res.status !== StatusCodes.OK) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const getUser = async (userId: number): Promise<UserType> => {
  const res = await axiosWithInterceptors.get(`/users/${userId}`);
  if (res.status !== StatusCodes.OK) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const getUsers = async ({
  q,
  paging,
}: {
  q: string;
  paging: PageParamsType;
}): Promise<ListWithPagingType<UserType>> => {
  const { cursor, size, order } = paging;
  const res = await axios.get(`/users`, {
    params: { q, cursor, size, order },
  });
  if (res.status !== StatusCodes.OK) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const getUserStatus = async (userId: number): Promise<UserType> => {
  const res = await axiosWithInterceptors.get(`/users/${userId}/status`);
  if (res.status !== StatusCodes.OK) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const getAchievements = async ({
  userId,
  paging,
}: {
  userId: number;
  paging: PageParamsType;
}): Promise<ListWithPagingType<AchievementType>> => {
  const { cursor, size, order } = paging;
  const res = await axiosWithInterceptors.get(`/users/${userId}/achievements`, {
    params: { cursor, size, order },
  });
  if (res.status !== StatusCodes.OK) {
    throw new Error(res.statusText);
  }
  return res.data;
};

/** Setting */

export const updateMyProfile = async (nickname: string): Promise<UserType> => {
  const res = await axiosWithInterceptors.patch(`/my/settings`, { nickname });
  if (res.status !== StatusCodes.OK) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const deleteMyProfileImage = async () => {
  const res = await axiosWithInterceptors.delete(`/my/profile-image`);
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

export const updateMyProfileImage = async (imageFile: File) => {
  const formData = new FormData();
  formData.append('file', imageFile);
  const res = await axiosWithInterceptors.post(`/my/profile-image`, formData, {
    headers: {
      'Content-Type': 'multipart/form-data',
    },
  });
  if (res.status !== StatusCodes.CREATED) {
    throw new Error(res.statusText);
  }
  return res;
};

export const checkProfile = async ({
  nickname,
}: {
  nickname: string;
}): Promise<{ nickname: boolean }> => {
  const res = await axios.post(`/my/settings/check`, { nickname });
  if (res.status !== StatusCodes.OK) {
    throw new Error(res.statusText);
  }
  return res.data;
};

/** Block */

export const getMyBlocks = async (
  paging: PageParamsType
): Promise<ListWithPagingType<UserType>> => {
  const { cursor, size, order } = paging;
  const res = await axiosWithInterceptors.get(`/my/blocks`, {
    params: { cursor, size, order },
  });
  if (res.status !== StatusCodes.OK) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const unblockUser = async (userId: number) => {
  const res = await axiosWithInterceptors.delete(`/my/blocks/${userId}`);
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

export const blockUser = async (userId: number) => {
  const res = await axiosWithInterceptors.put(`/my/blocks/${userId}`);
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

/** Follow */

export const getMyFollowings = async (
  paging: PageParamsType
): Promise<ListWithPagingType<UserType>> => {
  const { cursor, size, order } = paging;
  const res = await axiosWithInterceptors.get(`/my/following`, {
    params: { cursor, size, order },
  });
  if (res.status !== StatusCodes.OK) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const unfollowUser = async (userId: number) => {
  const res = await axiosWithInterceptors.delete(`/my/following/${userId}`);
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

export const followUser = async (userId: number) => {
  const res = await axiosWithInterceptors.put(`/my/following/${userId}`);
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

/** Channel */

export const getChannels = async ({
  enter,
  kind,
  paging,
}: {
  enter?: string;
  kind?: string[];
  paging: PageParamsType;
}): Promise<ListWithPagingType<ChannelType>> => {
  const { cursor, size, order } = paging;
  const res = await axiosWithInterceptors.get(`/channels/`, {
    params: { enter, kind, cursor, size, order },
  });

  if (res.status !== StatusCodes.OK) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const getChannel = async (channelId: number): Promise<ChannelType> => {
  const res = await axiosWithInterceptors.get(`/channels/${channelId}`);
  if (res.status !== StatusCodes.OK) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const createChannel = async (
  channelForm: ChannelFormType
): Promise<ChannelType> => {
  const res = await axios.post(`/channels`, channelForm);
  if (res.status !== StatusCodes.CREATED) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const joinChannel = async (channelForm: ChannelFormType) => {
  const res = await axiosWithInterceptors.put(`/channels/${channelForm.id}`, {
    password: channelForm.password,
  });
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

export const getDmChannel = async (userId: number): Promise<ChannelType> => {
  const res = await axiosWithInterceptors.get(`/dms/${userId}`);
  if (res.status !== StatusCodes.OK) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const getMessages = async ({
  channelId,
  paging,
}: {
  channelId: number;
  paging: PageParamsType;
}): Promise<ListWithPagingType<MessageType>> => {
  const { cursor, size, order } = paging;
  const res = await axiosWithInterceptors.get(
    `/channels/${channelId}/message`,
    {
      params: { cursor, size, order },
    }
  );
  if (res.status !== StatusCodes.OK) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const sendMessage = async ({
  channelId,
  message,
}: {
  channelId: number;
  message: string;
}) => {
  const res = await axiosWithInterceptors.post(
    `/channels/${channelId}/message`,
    {
      text: message,
    }
  );
  if (res.status !== StatusCodes.CREATED) {
    throw new Error(res.statusText);
  }
  return res;
};

export const updateChannelSetting = async (channelForm: ChannelFormType) => {
  const res = await axiosWithInterceptors.patch(`/channels/${channelForm.id}`, {
    title: channelForm.title,
    password: channelForm.password,
  });
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

export const inviteUserToChannel = async ({
  channelId,
  userIds,
}: {
  channelId: number;
  userIds: number[];
}) => {
  const res = await axiosWithInterceptors.put(`/channels/${channelId}/users`, {
    userIds,
  });
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

export const updateChannelUserRole = async ({
  channelId,
  userId,
  role,
}: {
  channelId: number;
  userId: number;
  role: ChannelUserRoleType;
}) => {
  const res = await axiosWithInterceptors.patch(
    `/channels/${channelId}/users/${userId}/role`,
    { role }
  );
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

export const updateChannelUserStatus = async ({
  channelId,
  userId,
  status,
}: {
  channelId: number;
  userId: number;
  status: ChannelUserStatusType;
}) => {
  const res = await axiosWithInterceptors.patch(
    `/channels/${channelId}/users/${userId}/status`,
    { status }
  );
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

export const leaveChannel = async (channelId: number) => {
  const res = await axiosWithInterceptors.delete(
    `/channels/${channelId}/users`
  );
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

/** Game */

export const getGame = async (gameId: number): Promise<GameType> => {
  const res = await axiosWithInterceptors.get(`/games/${gameId}`);
  if (res.status !== StatusCodes.OK) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const getGames = async (
  paging: PageParamsType
): Promise<ListWithPagingType<GameType>> => {
  const { cursor, size, order } = paging;
  const res = await axiosWithInterceptors.get(`/games`, {
    params: { cursor, size, order },
  });
  if (res.status !== StatusCodes.OK) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const waitGame = async (isLadder: boolean): Promise<AxiosResponse> => {
  const res = await axiosWithInterceptors.post(`/games/queue`, { isLadder });
  if (res.status !== StatusCodes.CREATED) {
    throw new Error(res.statusText);
  }
  return res;
};

export const cancelWaitingGame = async () => {
  const res = await axiosWithInterceptors.delete(`/games/queue`);
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

export const inviteGame = async (userId: number) => {
  const res = await axiosWithInterceptors.post(`/games/invite`, { userId });
  if (res.status !== StatusCodes.CREATED) {
    throw new Error(res.statusText);
  }
  return res;
};

export const cancelInvitingGame = async () => {
  const res = await axiosWithInterceptors.delete(`/games/invite`);
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

export const replyGameInvitation = async ({
  gameId,
  isAccepted,
}: {
  gameId: number;
  isAccepted: boolean;
}) => {
  const res = await axiosWithInterceptors.patch(`/games/${gameId}/invite`, {
    isAccepted,
  });
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

export const getGameSetting = async (gameId: number): Promise<GameType> => {
  const res = await axiosWithInterceptors.get(`/games/${gameId}/setting`);
  if (res.status !== StatusCodes.OK) {
    throw new Error(res.statusText);
  }
  return res.data;
};

export const updateGameSetting = async ({ id, mode, theme }: GameType) => {
  const res = await axiosWithInterceptors.patch(`/games/${id}/setting`, {
    mode,
    theme,
  });
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

export const startGame = async (gameId: number) => {
  const res = await axiosWithInterceptors.patch(`/games/${gameId}/play`);
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

export const joinGameLive = async (gameId: number) => {
  const res = await axiosWithInterceptors.patch(`/games/${gameId}/users`);
  if (res.status !== StatusCodes.NO_CONTENT) {
    throw new Error(res.statusText);
  }
  return res;
};

export const getGameHistories = async ({
  userId,
  paging,
}: {
  userId: number;
  paging: PageParamsType;
}): Promise<ListWithPagingType<GameHistoryType>> => {
  const { cursor, size, order } = paging;
  const res = await axiosWithInterceptors.get(`/games/users/${userId}`, {
    params: { cursor, size, order },
  });
  if (res.status !== 200) {
    throw new Error(res.statusText);
  }
  return res.data;
};
