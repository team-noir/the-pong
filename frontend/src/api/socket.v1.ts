import { socket } from 'contexts/socket';
import {
  AchievementType,
  GameResultType,
  MessageType,
  NoticeType,
  UserType,
} from 'types';
import SOCKET_EVENTS from 'constants/socketEvents';

export const onConnect = () => {
  socket.on(SOCKET_EVENTS.CONNECT, () => console.info('socket connected'));
};

export const onDisconnect = () => {
  socket.on(SOCKET_EVENTS.DISCONNECT, () =>
    console.info('socket disconnected')
  );
};

/** Game */

export const onPing = () => {
  socket.on(SOCKET_EVENTS.GAME.PING, () => emitPong());
};

export const onQueue = (
  callback: (data: { text: string; gameId?: number }) => void
) => {
  socket.on(SOCKET_EVENTS.GAME.QUEUE, callback);
};

export const onGameInvite = (
  callback: (data: { text: string; user?: UserType; gameId?: number }) => void
) => {
  socket.on(SOCKET_EVENTS.GAME.INVITE, callback);
};

export const onGameSetting = (
  callback: (data: { text: string; mode?: number; theme?: number }) => void
) => {
  socket.on(SOCKET_EVENTS.GAME.SETTING, callback);
};

export const onGameStart = (callback: () => void) => {
  socket.on(SOCKET_EVENTS.GAME.START, callback);
};

export const onGameViewer = (
  callback: (data: { viewerCount: number }) => void
) => {
  socket.on(SOCKET_EVENTS.GAME.VIEWER, callback);
};

export const onRoundOver = (
  callback: (data: { winnerId: number; score: number }) => void
) => {
  socket.on(SOCKET_EVENTS.GAME.ROUND_OVER, callback);
};

export const onGameOver = (callback: (data: GameResultType) => void) => {
  socket.on(SOCKET_EVENTS.GAME.GAME_OVER, callback);
};

export const onAchievment = (callback: (data: AchievementType) => void) => {
  socket.on(SOCKET_EVENTS.GAME.ACHIEVMENT, callback);
};

export const emitPong = () => {
  socket.emit(SOCKET_EVENTS.GAME.PONG);
};

export const emitRoundOver = (playerId: number) => {
  socket.emit(SOCKET_EVENTS.GAME.ROUND_OVER, { winnerId: playerId });
};

/** Game RTC */

export const onRtcInit = (callback: (data: { userId: number }) => void) => {
  socket.on(SOCKET_EVENTS.GAME.RTC.INIT, callback);
};

export const onRtcGetCandidate = (
  callback: (data: {
    candidate: RTCIceCandidateInit;
    candidateSendUserId: number;
  }) => void
) => {
  socket.on(SOCKET_EVENTS.GAME.RTC.GET_CANDIDATE, callback);
};

export const onRtcGetAnswer = (
  callback: (data: {
    sdp: RTCSessionDescription;
    answerSendUserId: number;
  }) => void
) => {
  socket.on(SOCKET_EVENTS.GAME.RTC.GET_ANSWER, callback);
};

export const onRtcGetOffer = (
  callback: (data: {
    sdp: RTCSessionDescription;
    offerSendUserId: number;
  }) => void
) => {
  socket.on(SOCKET_EVENTS.GAME.RTC.GET_OFFER, callback);
};

export const emitRtcCandidate = (
  candidate: RTCIceCandidate,
  candidateSendUserId: number,
  candidateReceiveUserId: number
) => {
  socket.emit(SOCKET_EVENTS.GAME.RTC.CANDIDATE, {
    candidate,
    candidateSendUserId,
    candidateReceiveUserId,
  });
};

export const emitRtcOffer = (
  sdp: RTCSessionDescriptionInit,
  offerSendUserId: number,
  offerReceiveUserId: number
) => {
  socket.emit(SOCKET_EVENTS.GAME.RTC.OFFER, {
    sdp,
    offerSendUserId,
    offerReceiveUserId,
  });
};

export const emitRtcAnswer = (
  sdp: RTCSessionDescriptionInit,
  answerSendUserId: number,
  answerReceiveUserId: number
) => {
  socket.emit(SOCKET_EVENTS.GAME.RTC.ANSWER, {
    sdp,
    answerSendUserId,
    answerReceiveUserId,
  });
};

export const emitConnected = (userId: number) => {
  socket.emit(SOCKET_EVENTS.GAME.RTC.CONNECTED, { userId });
};

/** Channel */

export const onMessage = (callback: (data: MessageType) => void) => {
  socket.on(SOCKET_EVENTS.CHANNEL.MESSAGE, callback);
};

export const onNotice = (callback: (data: NoticeType) => void) => {
  socket.on(SOCKET_EVENTS.CHANNEL.NOTICE, callback);
};

export const onUser = (callback: () => void) => {
  socket.on(SOCKET_EVENTS.CHANNEL.USER, callback);
};
