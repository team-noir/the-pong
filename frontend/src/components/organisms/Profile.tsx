import React from 'react';
import { UserType } from 'types/userType';
import styles from 'assets/styles/Profile.module.css';

interface Props extends UserType {
  nickname: string;
  profileImageUrl: string;
}

export default function Profile({ id, nickname, profileImageUrl }: Props) {
  return (
    <>
      <h1>Profile</h1>
      <img src={profileImageUrl} alt="profile image" className={styles.img} />
      <p data-testid={id}>{nickname}</p>
    </>
  );
}
