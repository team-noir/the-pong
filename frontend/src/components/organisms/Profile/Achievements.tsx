import { useInfiniteQuery } from '@tanstack/react-query';
import { AxiosError } from 'axios';
import { getAchievements } from 'api/rest.v1';
import InfiniteScroll from 'react-infinite-scroll-component';
import Spinner from 'components/atoms/Spinner';
import { AchievementType } from 'types';
import QUERY_KEYS from 'constants/queryKeys';

interface Props {
  userId: number;
}

export default function Achievements({ userId }: Props) {
  const { data, isFetching, fetchNextPage } = useInfiniteQuery({
    queryKey: [QUERY_KEYS.ACHIEVEMENTS, String(userId)],
    queryFn: ({ pageParam = null }) =>
      getAchievements({ userId, paging: { cursor: pageParam, size: 1 } }),
    getNextPageParam: ({ paging }) => paging.nextCursor,
    useErrorBoundary: (error: AxiosError) => {
      if (error && error.response?.status === 404) return false;
      return true;
    },
    suspense: false,
  });

  const achievements = data?.pages.flatMap((page) => page.data) ?? [];
  const hasMore = !!data?.pages[data.pages.length - 1].paging.nextCursor;

  return (
    <section className="section">
      <h2 className="section-title">업적</h2>
      {isFetching && !data ? (
        <Spinner className="flex justify-center mt-2 mb-8" />
      ) : achievements.length ? (
        <InfiniteScroll
          next={fetchNextPage}
          hasMore={hasMore}
          dataLength={achievements.length}
          loader={<Spinner className="flex justify-center pt-2 pb-8" />}
          scrollThreshold={0.7}
        >
          <AchievementList achievements={achievements} />
        </InfiniteScroll>
      ) : (
        <p className="block mt-4 text-center text-gray">
          게임을 플레이하고 업적을 달성해보세요.
        </p>
      )}
    </section>
  );
}

function AchievementList({
  achievements,
}: {
  achievements: AchievementType[];
}) {
  return (
    <ul className="flex flex-col">
      {achievements &&
        achievements.map((achievement) => {
          return (
            <AchievementItem key={achievement.id} achievement={achievement} />
          );
        })}
    </ul>
  );
}

function AchievementItem({ achievement }: { achievement: AchievementType }) {
  return (
    <li className="vh-center flex-col mb-4 py-4 px-6 text-center bg-gray-light text-text-dark border-gray-dark border-double border-4 border-r-0 border-l-0">
      <p className="text-xl mb-2">{achievement.title}</p>
      <p className="mb-2 italic">{achievement.description}</p>
      <p className="max-w-xs mb-2 text-sm">{achievement.condition}</p>
    </li>
  );
}
