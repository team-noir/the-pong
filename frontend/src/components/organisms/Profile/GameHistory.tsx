import { useQuery } from '@tanstack/react-query';
import { getGameHistories } from 'api/rest.v1';
import GameHistoryList from 'components/molecule/GameHistoryList';
import QUERY_KEYS from 'constants/queryKeys';

interface Props {
  userId: number;
}

export default function GameHistory({ userId }: Props) {
  const { data: histories } = useQuery({
    queryKey: [QUERY_KEYS.GAME_HISTORY, String(userId)],
    queryFn: () => getGameHistories(userId),
  });

  return (
    <section className="section">
      <h2 className="section-title">게임 히스토리</h2>
      {histories?.length ? (
        <GameHistoryList histories={histories} userId={userId} />
      ) : (
        <p className="block mt-4 text-center text-gray">
          게임 히스토리가 없습니다.
        </p>
      )}
    </section>
  );
}
