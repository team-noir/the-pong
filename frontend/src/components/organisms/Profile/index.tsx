import { Link, useNavigate } from 'react-router-dom';
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import * as api from 'api/api.v1';
import { useUser } from 'hooks/useStore';
import Button from 'components/atoms/Button';
import ProfileImage from 'components/atoms/ProfileImage';
import Achievements from 'components/organisms/Profile/Achievements';
import GameHistory from 'components/organisms/Profile/GameHistory';
import { UserType } from 'types';

interface Props {
  userId: number;
}

export default function Profile({ userId }: Props) {
  const myUserId = useUser((state) => state.id);
  const queryClient = useQueryClient();
  const navigate = useNavigate();

  const queryKey = ['profile', userId];
  const isMyPage = userId === myUserId;

  const { data: user } = useQuery({
    queryKey,
    queryFn: () => api.getUser(Number(userId)),
  });

  const onMutate = async (newUser: UserType | undefined) => {
    await queryClient.cancelQueries(queryKey);
    const prevUser = user;
    queryClient.setQueryData(queryKey, newUser);
    return { prevUser };
  };
  const onError = (
    _error: Error,
    _variables: number,
    context: { prevUser: UserType | undefined } | undefined
  ) => queryClient.setQueryData(queryKey, context?.prevUser);
  const onSettled = () => queryClient.invalidateQueries(queryKey);

  const followUserMutation = useMutation({
    mutationFn: api.followUser,
    onMutate: () => onMutate(user && { ...user, isFollowedByMyself: true }),
    onError,
    onSettled,
  });

  const unfollowUserMutation = useMutation({
    mutationFn: api.unfollowUser,
    onMutate: () => onMutate(user && { ...user, isFollowedByMyself: false }),
    onError,
    onSettled,
  });

  const blockUserMutation = useMutation({
    mutationFn: api.blockUser,
    onMutate: () => onMutate(user && { ...user, isBlockedByMyself: true }),
    onError,
    onSettled,
  });

  const unblockUserMutation = useMutation({
    mutationFn: api.unblockUser,
    onMutate: () => onMutate(user && { ...user, isBlockedByMyself: false }),
    onError,
    onSettled,
  });

  const getDmChannelMutation = useMutation({
    mutationFn: api.getDmChannel,
    onSuccess: (data) => navigate(`/channel/${data.id}`),
  });

  return (
    <>
      <section className="section">
        {user && (
          <div className="flex flex-col items-center">
            <ProfileImage userId={userId} alt="profile image" size={192} />
            <div className="mb-8 text-center">
              <p
                data-testid={userId}
                className="text-2xl font-semibold mt-4 mb-2"
              >
                {user.nickname}
              </p>
              <span className="badge">Lv. {user.level}</span>
            </div>
            {isMyPage && (
              <Link to="/setting/profile" className="button primary small">
                프로필 수정하기
              </Link>
            )}
            {!isMyPage && (
              <div className="flex flex-col items-center">
                <div className="inline-flex gap-4 mb-4">
                  {user.isFollowedByMyself ? (
                    <Button
                      onClick={() => unfollowUserMutation.mutate(userId)}
                      secondary
                      className="w-36"
                    >
                      언팔로우
                    </Button>
                  ) : (
                    <Button
                      onClick={() => followUserMutation.mutate(userId)}
                      primary
                      className="w-36"
                    >
                      팔로우하기
                    </Button>
                  )}
                  {!user.isBlockedByMyself && (
                    <Button
                      onClick={() => getDmChannelMutation.mutate(userId)}
                      primary
                      className="w-36"
                    >
                      메시지 보내기
                    </Button>
                  )}
                </div>
                {user.isBlockedByMyself ? (
                  <Button
                    onClick={() => unblockUserMutation.mutate(userId)}
                    linkStyle
                    className="text-red"
                    size="small"
                  >
                    차단 해제
                  </Button>
                ) : (
                  <Button
                    onClick={() => blockUserMutation.mutate(userId)}
                    linkStyle
                    className="text-red"
                    size="small"
                  >
                    차단하기
                  </Button>
                )}
              </div>
            )}
          </div>
        )}
      </section>
      <Achievements userId={Number(userId)} />
      <GameHistory userId={Number(userId)} />
    </>
  );
}
