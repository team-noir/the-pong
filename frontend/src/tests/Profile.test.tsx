import { render, screen } from '@testing-library/react';
import { ProfileUserType } from 'api/api.v1';
import Profile from 'components/organisms/Profile';

const profileUser: ProfileUserType = {
  id: 1,
  nickname: `user1's nickname`,
  rank: 0,
  achievements: [],
  games: [],
  isFollowedByMyself: false,
  isBlockedByMyself: false,
};

const mockHandler = (userId: number, message: string) =>
  console.log(`${userId} ${message}}`);

describe('Component - Profile 렌더링', () => {
  test('닉네임과 프로필 사진을 보여준다', () => {
    render(
      <Profile
        user={profileUser}
        myId={'2'}
        onClickFollow={(userId: number) => mockHandler(userId, 'followed')}
        onClickUnfollow={(userId: number) => mockHandler(userId, 'unfollowed')}
        onClickBlock={(userId: number) => mockHandler(userId, 'blocked')}
        onClickUnblock={(userId: number) => mockHandler(userId, 'unblocked')}
        onClickDm={(userId: number) => mockHandler(userId, 'dm')}
      />
    );
    screen.findByText(profileUser.nickname);
    screen.findByAltText('profile image');
    screen.findByText('팔로우하기');
  });
});
