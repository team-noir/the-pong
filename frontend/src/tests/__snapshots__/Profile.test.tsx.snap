// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile /> matches snapshot 1`] = `
<div>
  <h1>
    Profile
  </h1>
  <img
    alt="profile image"
    class="img"
    src="https://placekitten.com/800/800"
  />
  <p>
    nickname
  </p>
</div>
`;
